### ISSUE TITLE
Request: `no-extra-parens` option to ignore `&&`/`||`

### ISSUE THREAD

**Issue by ljharb** [eslint/eslint#3065]
Although the rules for precedence of `&&` and `||` are clear, using parentheses to visually group these is often very helpful to ensure programmer intentions. ie: `a || (b && c)` even though `a || b && c` is the same.

Most of the time, I don't want extra parens - but around selection operators, I very much do. Can we add an option to this rule so it ignores `&&`/`||` operators? (for stylistic reasons only)

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by sindresorhus** [comment]
:+1: I do this too for the sake of readability.

**Comment by nzakas** [comment]
We'd need some generic way to do this, since these are all BinaryExpression nodes, but this rule applies not just to those types of nodes. Suggestions?

**Comment by michaelficarra** [comment]
@sindresorhus Then why don't you just disable this rule? What are you using it for at that point?

**Comment by ljharb** [comment]
@michaelficarra For every other expression value? I don't want extra parens around strings, ternary conditionals - the only thing I want to allow 1 extra paren around is binary expressions. Someone could do `var x = (3);` and that shouldn't be allowed, to use a contrived example.

@nzakas What about using the esprima node type in a list of exceptions? Kind of like https://github.com/jscs-dev/node-jscs/blob/566d9e48e95b480e78217445b2bf3d102b2151b6/lib/rules/disallow-node-types.js

Using node types might also allow for an extensible way to do what @sindresorhus is suggesting, assuming there's a node type for math operations.

**Comment by nzakas** [comment]
The nodetype alone wouldn't be enough, though, because you want to focus in on the operators being used.

**Comment by sindresorhus** [comment]
> What are you using it for at that point?

Lots of things. Like preventing the following (which I often see in PRs):

```js
var foo = (x ? 'foo' : 'bar');

if ((foo === true)) {}

if ((typeof foo) === 'string') {}

return (foo || bar);

var foo = (45 + 45);
```

**Comment by michaelficarra** [comment]
Okay, then how about we add an option to allow binary expressions within other binary expressions to be parenthesised, regardless of whether they're necessary?

**Comment by ljharb** [comment]
At first glance that seems both reasonable and like it would entirely cover my use cases.

**Comment by sindresorhus** [comment]
:+1:

**Comment by nzakas** [comment]
Sounds good. Suggestions for the option name?

**Comment by ljharb** [comment]
`ignoreNestedBinaryExpressions`?

**Comment by nzakas** [comment]
Hmm, may need a bit of thought considering we have "all" and "functions" as options now, so this would only apply in the "all" case

**Comment by ljharb** [comment]
It could still be a separate option that was just a noop in "functions" mode - alternatively, all/functions/allButNestedBinaryExpressions (but that seems gnarly)

**Comment by nzakas** [comment]
Yeah, not a big fan of no-ops - rule configuration is confusing enough as it is without throwing things in that sometimes do something and sometimes don't. Need to think a bit more about this.

Other suggestions?

**Comment by ljharb** [comment]
Fair point. The problem with "all" and "functions" is that it kind of prevents any further modification to the rule structure.

I'd thought of `{ "ignoreNestedBinaryExpressions": false, "ignoreFunctions": false }` to mean "all", but then to get "functions" you'd need `{ "ignoreNestedBinaryExpressions": true, "ignoreFunctions": false, "ignoreEverythingElse": true }` which is funky.

**Comment by nzakas** [comment]
Yup, welcome to my life. :)

**Comment by hectorcascos** [comment]
What about `exceptNestedBinaryExpressions`?

I think here "all" is implicit

**Comment by ljharb** [comment]
@hectorcascos so `{ "exceptNestedBinaryExpressions": true, "all": true }` would do what we expect, but how would you combine `{ "exceptNestedBinaryExpressions": true }` with something for "functions"?

**Comment by hectorcascos** [comment]
I'm suggesting to add a separate option.

No-ops can lead to confusion, but using `exceptNestedBinaryExpressions` is clear what it does.

**Comment by ljharb** [comment]
ah, you mean the three choices of "all", "functions", or "exceptNestedBinaryExpressions". That seems sensible to me - @nzakas, would that work for you/other collaborators?

**Comment by nzakas** [comment]
That doesn't quite fit with our naming conventions. We generally don't do camelcase for named string options, nor is it clear that `exceptNestedBinaryExpressions` implies `all`. So, I don't think we're quite there yet.

**Comment by ljharb** [comment]
`allExceptNestedBinaryExpressions` or something might be better, but that doesn't handle the camelcasing concern :-/

**Comment by tsilenzio** [comment]
What about using `nestedBinaryExpressions` (or `nestedBinaryExpression`) within an object, similar to `conditionalAssign`? Like the following:
```javascript
"no-extra-parens": [2, "all", {"conditionalAssign": true, "nestedBinaryExpressions": true}]
```

**Comment by ljharb** [comment]
@tsilenzio wouldn't that make absence of the "nestedBinaryExpressions" option the same as "true"? Typically it's not a good idea to make absence default to a truthy value, for readability.

**Comment by tsilenzio** [comment]
Doesn't the flag `conditionalAssign` follow the same pattern as defined on: http://eslint.org/docs/2.0.0/rules/no-extra-parens.html?

**Comment by ljharb** [comment]
That's a fair point. @nzakas, do you think `nestedBinaryExpressions` would be an acceptable option name?

**Comment by nzakas** [comment]
Would `nestedBinaryExceptions` only apply when using `"all"`?

**Comment by tsilenzio** [comment]
Yes, similar to `conditionalAssign`.

**Comment by nzakas** [comment]
Sounds fine to me. Does someone want to submit a PR?

**Comment by ljharb** [comment]
@ilyavolodin thank you!

**Comment by spadgos** [comment]
:bow:

**Comment by ilyavolodin** [comment]
Re-opening this issue, since the commit has been reversed for a patch release. @spadgos I'm very sorry about this, could you re-open your PR please?

**Comment by nzakas** [comment]
@ilyavolodin you can just revert the revert locally rather than doing another PR

**Comment by obedm503** [comment]
is there a way for eslint to also ignore extra parens in the conditional of a ternary expression, so that
```js
// valid
const a = (b || c) ? 5 : 20;
const r = (n * 2) ? a : b;
// invalid
const a = b || c ? 5 : 20;
const r = n * 2 ? a : b;
```
is valid. but it should kick in when there are no binary operators
```js
// valid
const a = b ? 5 : 20;
// invalid
const a = (b) ? 5 : 20;
```
if already possible, can someone point me to the correct configuration
